File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    boolean dfs(int idx, ArrayList<ArrayList<Integer>> adjList, int[] visit, Stack<Integer> S){
        visit[idx] = 1;
        
        boolean flag = true;
        for(int to : adjList.get(idx)){
            if(visit[to]==0){
                if(!dfs(to, adjList, visit, S))
                    return false;
            }
            else if(visit[to]==1)
                return false;
        }
        
        S.push(idx);
        visit[idx] = 2;
        return flag;
    }
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        ArrayList<ArrayList<Integer>> adjList = new ArrayList<ArrayList<Integer>>();
        int[] visit = new int[numCourses];
        for(int i=0; i<numCourses; i++){
            adjList.add(new ArrayList<>());
        }
        
        for(int[] arr : prerequisites){
            adjList.get(arr[1]).add(arr[0]);
        }
        
        var S = new Stack<Integer>();
        for(int i=0; i<numCourses; i++){
            if(visit[i]==0){
                boolean flag = dfs(i, adjList, visit, S);
                if(!flag)
                    return new int[0];
            }
        }
        
        int[] sol = new int[numCourses];
        for(int i=0; i<numCourses; i++)
            sol[i] = S.pop();
        return sol;
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    bool dfs(int cur, unordered_map<int, vector<int>>& m, unordered_set<int>& cycle, unordered_set<int>& visit, vector<int>& sol){
        if(cycle.find(cur)!=cycle.end())
            return false;
        if(visit.find(cur)!=visit.end())
            return true;
        
        cycle.insert(cur);
        for(int v : m[cur]){
            if(!dfs(v, m, cycle, visit, sol))
                return false;
        }
        
        m[cur].clear();
        cycle.erase(cur);
        visit.insert(cur);
        sol.push_back(cur);
        return true;
    }
    
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> m;
        for(auto& v : prerequisites) {
            m[v[1]].push_back(v[0]);
        }
        
        unordered_set<int> cycle;
        unordered_set<int> visit;
        vector<int> sol;
        for(int i=0; i<numCourses; i++){
            if(!dfs(i, m, cycle, visit, sol))
                return {};
        }
        reverse(sol.begin(), sol.end());
        return sol;
    }
};